### ISSUE TITLE
Test failure: System.Text.Json.Serialization.Tests.CacheTests/MultipleThreads

### ISSUE THREAD

**Issue by AriNuer** [dotnet/corefx#38802]
Test `System.Text.Json.Serialization.Tests.CacheTests/MultipleThreads` has failed.

Message :
```
System.AggregateException : One or more errors occurred. (The JSON value could not be converted to System.String. Path: $.MyStringToStringImmutableSortedDict | LineNumber: 0 | BytePositionInLine: 2707.)
---- System.Text.Json.JsonException : The JSON value could not be converted to System.String. Path: $.MyStringToStringImmutableSortedDict | LineNumber: 0 | BytePositionInLine: 2707.
```
Stack Trace :
```
   at System.Threading.Tasks.Task.WaitAllCore(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 4801
   at System.Text.Json.Serialization.Tests.CacheTests.MultipleThreads() in /_/src/System.Text.Json/tests/Serialization/CacheTests.cs:line 49
----- Inner Stack Trace -----
   at System.Text.Json.ThrowHelper.ThrowJsonException(String message, Utf8JsonReader& reader, String path, Exception innerException) in /_/src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs:line 136
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType, Utf8JsonReader& reader, String path, Exception innerException) in /_/src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs:line 35
   at System.Text.Json.JsonPropertyInfoNotNullable`4.OnReadEnumerable(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs:line 62
   at System.Text.Json.JsonPropertyInfo.ReadEnumerable(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs:line 355
   at System.Text.Json.JsonPropertyInfo.Read(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs:line 347
   at System.Text.Json.JsonSerializer.HandleValue(JsonTokenType tokenType, JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleValue.cs:line 29
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs:line 48
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Helpers.cs:line 24
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 74
   at System.Text.Json.JsonSerializer.Parse[TValue](String json, JsonSerializerOptions options) in /_/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 31
   at System.Text.Json.Serialization.Tests.CacheTests.<MultipleThreads>g__DeserializeObjectNormal|1_1() in /_/src/System.Text.Json/tests/Serialization/CacheTests.cs:line 26
   at System.Text.Json.Serialization.Tests.CacheTests.<>c.<MultipleThreads>b__1_4() in /_/src/System.Text.Json/tests/Serialization/CacheTests.cs:line 42
   at System.Threading.Tasks.Task.InnerInvoke() in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2449
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2428
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:line 289
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:line 314
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2389
```
Build: -[20190623.30](https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190623.30)(Master)

Failing configurations:

- OSX.1013.Amd64.Open-x64-Release

Details:
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190623.30/workItem/System.Text.Json.Tests/analysis/xunit/System.Text.Json.Serialization.Tests.CacheTests~2FMultipleThreads

**Comment by ahsonkhan** [comment]
cc @steveharter, @JeremyKuhne

**Comment by steveharter** [comment]
Please add additional hits. The code was refactored and this may be fixed.

The regression was likely [here](https://github.com/dotnet/corefx/pull/38705/files#diff-d94322d11e50b60b48c128b06c14bf1dR69) to address this [issue](https://github.com/dotnet/corefx/issues/38528). This does a HashSet.CopyTo which on multiple threads could cause overlapping on the array.

The code was recently [refactored ](https://github.com/dotnet/corefx/pull/38902/files#diff-d94322d11e50b60b48c128b06c14bf1dR58) to overwrite the array variable instead of appending.

**Comment by steveharter** [comment]
I will keep this issue open until I can perform additional local stress testing, add additional tests, and review the existing code (which should have already fixed this issue).

**Comment by ahsonkhan** [comment]
Re-opening until its fixed in release/3.0.

**Comment by ahsonkhan** [comment]
Fixed in 3.0 with https://github.com/dotnet/corefx/pull/40356.
